Fix changing payment to saved payment for subscriptions order

shendy-a8c opened the pull request:
Fixes #1185

#### Changes proposed in this Pull Request

* Override `WC_Payment_Gateway_WCPay::process_payment()` in `WC_Payment_Gateway_WCPay_Subscriptions_Compat::process_payment()` to handle specifically payment method change for subscription order, because:
  * [On caught exception, it updates the order's status to failed](https://github.com/Automattic/woocommerce-payments/blob/trunk/includes/class-wc-payment-gateway-wcpay.php#L470).
  * Taking out [this return statement](https://github.com/Automattic/woocommerce-payments/blob/62516cdcee37993cae25999c6256d5c574e53124/includes/class-wc-payment-gateway-wcpay.php#L571-L574) per [suggested here](https://github.com/Automattic/woocommerce-payments/pull/1148#issuecomment-750410406) is not enough because [stock will be reduced and cart will be emptied on success](https://github.com/Automattic/woocommerce-payments/blob/62516cdcee37993cae25999c6256d5c574e53124/includes/class-wc-payment-gateway-wcpay.php#L711-L714).
  * `WC_Payment_Gateway_WCPay` shouldn't know the nuts and bolts of subscription, so doing something like "if changing payment method for subscription, do not reduce stock", for example, IMO breaks the abstraction.

#### Testing instructions

Shamelessly copied from [1185](https://github.com/Automattic/woocommerce-payments/issues/1185).
1. Purchase subscription product.
2. Go to My account > Add one more payment method.
3. Go to My account > Subscription > Purchased subscription.
4. Try to change payment method to saved on step 2.
5. Success message is displayed. With `trunk`, payment method is not changed. With this branch, payment method should be changed.

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

shendy-a8c committed:
Remove return statement that prevents add_token_to_order() being called in process_payment_for_order() when changing payment method for subscriptions order.

shendy-a8c committed:
Handles changing payment method for subscriptions order in WC_Payment_Gateway_WCPay_Subscriptions_Compat, overriding process_payment_for_order().

- Revert previous change that removed a return statement in WC_Payment_Gateway_WCPay::process_payment_for_order().

shendy-a8c committed:
- Fix for changing payment method subscriptions check.
- Call payment_complete() so order's status is updated correctly.

shendy-a8c committed:
- Add order note when payment method is changed.
- Handles intent when requires action (SCA).

shendy-a8c committed:
- DRY code block to upsert customer from user.
- Override process_payment() instead of process_payment_for_order().
- Handle failed intent.

shendy-a8c committed:
- Add doc comments.
- Fix linting issues.
- When changing payment method, always setup intent even though for saved payment method.
- Fix order notes.
- upsert_customer() returns $customer_id.

shendy-a8c committed:
Fix linting problem.

shendy-a8c committed:
Test updating subscription payment method using saved payment method.

shendy-a8c committed:
Changelog.

shendy-a8c commented:
I have questions:
1. When changing payment method for subs order, do we need to [update customer](https://github.com/Automattic/woocommerce-payments/pull/1199/files#diff-a5e8d94b5af9b7536f23cf436b4eef1fb76da3069f34a63b4cf35c46051e417bR97)? I can see that possibly this is not preferable because the billing information from subs order might be stale data, for example, after 1 year of active subs, user updates payment method for that subs order. The billing details on that subs order will be 1-year old data.

2. I want to show `Failed to change payment method to Credit Card ending in 1234 with the following message: <code>%s</code>.` [here](https://github.com/Automattic/woocommerce-payments/pull/1199/files#diff-a5e8d94b5af9b7536f23cf436b4eef1fb76da3069f34a63b4cf35c46051e417bR112) but not sure how to get last 4 digit from. Is that order note sufficient without the last 4 digit of credit card? Or should I include that information in the order note?
Do I get last 4 digit this way?
```php
$payment_method_id = WCPay\Payment_Information::get_payment_method_from_request( $request );
$payment_method = $this->payments_api_client->get_payment_method( $payment_method_id );
// $payment_method['card']['last4']
```

3. Currently, updating payment method to credit card that requires SCA like `4000002500003155` does not work. Any pointer?

shendy-a8c committed:
Fix linting problem.

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -513,6 +513,35 @@ protected function prepare_payment_information( $order ) {
+	protected function upsert_customer( $user, $name, $email ) {
+		if ( ! $user ) {
+			return;
+		}
```
Comparing to the original code, this if-statement looks like dead code. Any specific reason why we adding it?

shendy-a8c replied:
Ah you're right. The only possible source of $user is from this statement
```php
$order->get_user() ?? wp_get_current_user()
```
`$order->get_user()` alone can return false but `wp_get_current_user()` always return `WP_User`.

I'm removing this check.

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -59,14 +59,14 @@ class WC_Payment_Gateway_WCPay extends WC_Payment_Gateway_CC {
 	 * @var WC_Payments_Token_Service
 	 */
-	private $token_service;
+	protected $token_service;
```
Not sure why do we need to change the visibility here. Can we keep those fields private?

shendy-a8c replied:
This is necessary so that I can access it from `WC_Payment_Gateway_WCPay_Subscriptions_Compat`.

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -513,6 +513,35 @@ protected function prepare_payment_information( $order ) {
+	 *
+	 * @return string|null Customer id.
+	 */
+	protected function upsert_customer( $user, $name, $email ) {
```
And related to the comment below: we can add type hints for parameters here.

shendy-a8c replied:
OK. I also just realized `create_customer_for_user()` and `update_customer_for_user()` can throw an exception. Will handle that as well.

kalessil commented on includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php:
```diff
@@ -77,6 +77,112 @@ private function is_changing_payment_method_for_subscription() {
+		try {
+			$intent = $this->payments_api_client->create_and_confirm_setup_intent(
+				$payment_information->get_payment_method(),
+				$customer_id
```
Let's use `$this->upsert_customer( $user, $name, $email )` here  for better context scoping.

shendy-a8c replied:
Good idea. Will do.

kalessil commented on includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php:
```diff
@@ -77,6 +77,112 @@ private function is_changing_payment_method_for_subscription() {
+				'redirect' => sprintf(
+					'#wcpay-confirm-si:%s:%s:%s',
+					$order_id,
+					$client_secret,
```
Let's also inline and use `$intent['client_secret']` here.

shendy-a8c replied:
OK, sure, but I wonder why?

kalessil replied:
To reduce the amount of changes mainly. Less changes help me as a reviewer to better understand what the change about and often find ways to simplify a few lines.

kalessil commented on includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php:
```diff
@@ -77,6 +77,112 @@ private function is_changing_payment_method_for_subscription() {
+			];
+		}
+
+		// TODO handle when intent failed.
```
Is this relevant? If so, let's create a new ticket and link it here like 'TODO: ticket url'

kalessil commented on includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php:
```diff
@@ -77,6 +77,112 @@ private function is_changing_payment_method_for_subscription() {
+
+		$order->set_transaction_id( $intent_id );
+		$order->update_meta_data( '_intent_id', $intent_id );
+		$order->update_meta_data( '_charge_id', $charge_id );
```
`$charge_id` is always an empty string, am I right?

shendy-a8c committed:
- Remove dead check since upsert_customer() will never receive empty $user.
- Update upsert_customer()'s doc.
- Move variable assignment closer to context where it's used.

shendy-a8c committed:
Do not update customer using billing details from subscriptions order when changing payment for subscriptions order.
Remove upsert_customer() since now there is no need to DRY the code.

shendy-a8c committed:
Mention last 4 digit of the credit card in the order note when an exception is caught when trying to confirm setup intent.

shendy-a8c committed:
Only call setup intent if new payment method is not a saved payment.

shendy-a8c committed:
Handle failed intent.
Fix the way to get last 4 digit credit card from new unsaved payment.
Fix test

shendy-a8c commented:
After discussion with @vbelolapotkov, it becomes clear that this PR tries to solve things that are out of the scope of the issue. Closing it and creating a new PR.
